### PR TITLE
FIX service path persistence at subscription creation (along with some refactoring)

### PR DESCRIPTION
**Pull request by fgalan**
"No .test was harmed in the making of this PR" :)

Apart of fixing #2024 itself, this PR moves the SP defaulting logic to mongoBackend, thus allowing to simplify removing defaultServicePath() funtion. In addition, some refactors has been include to make centralize SP filtering functionality across mongoBackend in the fillQueryServicePath() function.

Finally, unit test code has been optimized, removing duplicated copies of the "mock" service path.

Test done to check that #2024 has been solved with this PR:

```
$  curl -v -H "Content-Type: application/json" -H "Fiware-Service: test_service_path" http://localhost:1026/v2/subscriptions -d '{"notification": {"callback": "http://localhost:1234", "attributes": ["temperature"]}, "expires": "2016-04-05T14:00:00.00Z", "subject": {"entities": [{"idPattern": ".*", "type": "room"}], "condition": {"attributes": ["temperature"]}}}' 
* Hostname was NOT found in DNS cache
*   Trying ::1...
* Connected to localhost (::1) port 1026 (#0)
> POST /v2/subscriptions HTTP/1.1
> User-Agent: curl/7.38.0
> Host: localhost:1026
> Accept: */*
> Content-Type: application/json
> Fiware-Service: test_service_path
> Content-Length: 234
> 
* upload completely sent off: 234 out of 234 bytes
< HTTP/1.1 201 Created
< Connection: Keep-Alive
< Content-Length: 0
< Location: /v2/subscriptions/570d2a54fd070d4aad115e79
< Fiware-Correlator: 732925b2-00d0-11e6-9a87-000c29173617
< Date: Tue, 12 Apr 2016 17:03:16 GMT
< 
* Connection #0 to host localhost left intact
```

mongo result:

```
> db.csubs.find({},{servicePath:1})
{ "_id" : ObjectId("570d2a54fd070d4aad115e79"), "servicePath" : "/#" }
```

No CNR changes needed (as #2024 is a consecuence of a change already included in CNR).

@kzangeli 
@crbrox

### CODE REVIEW

File: src/lib/rest/rest.cpp [hunk 1]
```diff
@@ -807,72 +815,6 @@ bool urlCheck(ConnectionInfo* ciP, const std::string& url)
...
-  if (strcasestr(url, "contextAvailabilitySubscriptions") != NULL)      return DEFAULT_SERVICE_PATH_QUERIES;
-  if (strcasestr(url, "contextSubscriptions") != NULL)                  return DEFAULT_SERVICE_PATH_QUERIES;
-
-  if (strcasecmp(method, "POST")   == 0)                                return DEFAULT_SERVICE_PATH_UPDATES;
-  if (strcasecmp(method, "PUT")    == 0)                                return DEFAULT_SERVICE_PATH_UPDATES;
-  if (strcasecmp(method, "DELETE") == 0)                                return DEFAULT_SERVICE_PATH_UPDATES;
-  if (strcasecmp(method, "GET")    == 0)                                return DEFAULT_SERVICE_PATH_QUERIES;
-  if (strcasecmp(method, "PATCH")  == 0)                                return DEFAULT_SERVICE_PATH_UPDATES;
-
-  return DEFAULT_SERVICE_PATH_UPDATES;
```
Comment (fgalan) [inline src/lib/rest/rest.cpp]: On a second thought, a CNR entry is needed. Note that with the defaultServicePath() approach POST /v2/subscription was incorrectly using "/" as its URL is not taken into account as a special case and the POST default applies.

Comment (fgalan) [inline src/lib/rest/rest.cpp]: Fixed in a1676de2870383872982fb02caab4a3d78d4688e (actually adding #2024 to an existing CNR entry).

File: CHANGES_NEXT_RELEASE [hunk 2]
```diff
@@ -8,6 +8,6 @@
 - Fix: csub cache at mongoBackend update context subscription logic wrongly setting expiration with current time, ignoring actual expires/duration in the request
 - Fix: GET /v2/types response payload, from key-map to array (#1832)
 - Fix: GET /v2/types response payload, now using array for attribute types (#1636)
-- Fix: using "/#" as default service path instead of "/" in subscriptiobns (#1961)
+- Fix: using "/#" as default service path instead of "/" in subscriptiobns (#1961, #2024)
```
Comment (kzangeli) [inline CHANGES_NEXT_RELEASE]: old typo: subscriptiobns

Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: Fixed in 2627a0f

File: test/unittests/mongoBackend/mongoNotifyContext_test.cpp [hunk 3]
```diff
@@ -47,7 +48,7 @@ extern void setMongoConnectionForUnitTest(DBClientBase*);
 *
 * servicePathV - empty service path vector used for these tests 
 */
-static const std::vector<std::string> servicePathV;
+//static const std::vector<std::string> servicePathV;
```
Comment (kzangeli) [inline test/unittests/mongoBackend/mongoNotifyContext_test.cpp]: If you no longer use it, just remove it, with comment header and all ...

File: test/unittests/mongoBackend/mongoNotifyContext_test.cpp [hunk 4]
```diff
@@ -498,7 +505,7 @@ TEST(mongoNotifyContextRequest, EntNAttr1)
     setTimer(timerMock);
```
Comment (kzangeli) [inline test/unittests/mongoBackend/mongoNotifyContext_test.cpp]: utInit just might prepare a timerMock ...

File: test/unittests/unittest.h [hunk 5]
```diff
@@ -62,14 +62,6 @@ extern std::map<std::string, bool> options;
...
-*/
-extern std::vector<std::string> servicePathV;
-
- 
-
-/* ****************************************************************************
-*
 * servicePathVector - 
 */
 extern std::vector<std::string> servicePathVector;
```
Comment (kzangeli) [inline test/unittests/unittest.h]: So we had two vars for the same purpose?  :-)

NTC

File: test/unittests/mongoBackend/mongoNotifyContext_test.cpp (hunk 3 above)
Comment (fgalan) [inline test/unittests/mongoBackend/mongoNotifyContext_test.cpp]: Seems to be a leftover. I'll remove.

File: test/unittests/mongoBackend/mongoNotifyContext_test.cpp (hunk 4 above)
Comment (kzangeli) [inline test/unittests/mongoBackend/mongoNotifyContext_test.cpp]: NTC

**Comment by kzangeli** [comment]
LGTM

**Comment by crbrox** [comment]
LGTM

**Pull request merged**